plugins/inputs/cloud_pubsub/pubsub.go

```diff
@@ -170,30 +177,32 @@ func (ps *PubSub) startReceiver(parentCtx context.Context) error {
 
 // onMessage handles parsing and adding a received message to the accumulator.
 func (ps *PubSub) onMessage(ctx context.Context, msg message) error {
+	defer msg.Ack()
 	if ps.MaxMessageLen > 0 && len(msg.Data()) > ps.MaxMessageLen {
-		msg.Ack()
 		return fmt.Errorf("message longer than max_message_len (%d > %d)", len(msg.Data()), ps.MaxMessageLen)
 	}
 
-	var data []byte
+	// This function is called concurrently, but the decoder cannot.
+	ps.decoderMutex.Lock()
+	data, err := ps.decoder.Decode(msg.Data())
+	ps.decoderMutex.Unlock()
```

Comment:
I took a look at this and I agree with @srebhan that it would be better not to modify the shared code in content_coding.go. Could you copy the buffer here instead of in the decoder function itself?

Reply:
Sounds good. I have moved the copy buffer code to just the pubsub plugins.